### DISCUSSION TITLE
Completely destroy Alpine component instance from memory

### DISCUSSION THREAD

**Discussion by josephmiclaus** [alpinejs/alpine#2391] [1. Help] [answered]
Is there a way to completely destroy an Alpine component's instance from memory?

Let's say I have an Alpine component "counter" which counts up every second using `setInterval`.
Also I have a script which removes the DOM element of the counter component after 5 seconds.
Even though the DOM element has been removed, the counter interval still runs.
Therefore, when the DOM element is removed, I use the Alpine function `destroy()`, which is automatically called.
There I clear the counter interval.
But now I would like to completely destroy the Alpine component instance from memory.
If I can completely destroy the instance from memory, I wouldn't need to execute `clearInterval()` anymore and I would keep the memory clean.

Is there a way to do this?

Maybe you can understand my issue better when written in code, see below:

```html
<div id="counter" x-data="counter" x-ref="counter">0</div>

<script type="text/javascript">
    setTimeout(function() {
        document.getElementById('counter').remove();
    }, 5000);
</script>
```

```js
Alpine.data('counter', () => ({
    counter: 0,
    counter_interval: null,

    init() {
        this.counter_interval = setInterval(() => {
            this.counter++;
            this.$refs.counter.innerText = this.counter;
        }, 1000);
    },

    destroy() {
        clearInterval(this.counter_interval);
        // how to destroy this instance from memory?
    },
}));
```

No harm wanted, but this would be nice. 😄

ANSWER by @SimoTod:
https://github.com/alpinejs/alpine/discussions/2391#discussioncomment-1673047

**Comment by SimoTod** [comment, marked as answer]
setInterval is just standard javascript, not Alpine. The code you have is already the correct way.

**Reply by josephmiclaus** [reply]
Yes, that is common knowledge. 😄 

The question is: how can we destroy the instance of the Alpine component from memory.

Because simply removing the DOM element of the Alpine component, doesn't destroy it's instance from memory.

**Reply by SimoTod** [reply]
If you remove the document from the dom, Alpine will release all effects so unless you have a dangling reference in your code, when the garbage collector runs it should remove it from memory. 

Can you make an example to show that it's not true?

That being said, removing the element from memory wouldn't clear setInterval (if you call `setInterval(() => console.log('tick'), 1000)`, it stays in memory even if you don't assign it to a variable).

**Reply by josephmiclaus** [reply]
What is interesting is that `clearInterval(this.counter_interval)` within `destroy()` doesn't seem to actually clear the interval.

It seems that `clearInterval(this.counter_interval)` only works within e.g. `init()`.

~~Any idea why clearing the interval is not possible within `destroy()`?~~
Any idea why I cannot access `this.counter_interval` within `destroy()`?

**Reply by josephmiclaus** [reply]
Okay I understand now better how Alpine works (regarding releasing all effects), thanks!

> That being said, removing the element from memory wouldn't clear setInterval (if you call setInterval(() => console.log('tick'), 1000), it stays in memory even if you don't assign it to a variable).

You're right! What I actually assign to the variable is the `intervalID` of the interval.

However it seems that I cannot access that variable within the `destroy()` function. Thus I cannot clear the interval within `destroy()`.

**Reply by josephmiclaus** [reply]
Is it maybe because all properties have already been cleared/removed by the garbage collector?

I don't seem to be able to access any properties within `destroy()`:

```js
Alpine.data('foo', () => ({
    test: 'bar',

    destroy() {
        // this outputs undefined
        console.debug(this.test);
    },
}));
```

**Reply by SimoTod** [reply]
right, I think it's the current implementation that makes destroy less useful.
https://github.com/alpinejs/alpine/blob/main/packages/alpinejs/src/directives/x-data.js#L36-L39
Basically, it removes the scope before adding destroy so when you use this in your destroy callback, it doesn't point to your variables anymore. I believe undo() should be called after destroy.

**Reply by SimoTod** [reply]
In the meantime, you can use something like
```javascript
Alpine.data('counter', () => {
  let counter_interval // Define your variable in the function scope
  return {
    counter: 0,

    init() {
      counter_interval = setInterval(() => {
        this.counter++
        if (this.counter === 10) this.$el.remove()  // You don't need an external setTimeout
      }, 1000);
    },

    destroy() {
      clearInterval(counter_interval)
      delete counter_interval // Just in case
    },
  }
});
```

**Reply by josephmiclaus** [reply]
Okay gotcha! Thanks! 

So I won't be using `destroy()` to clear the interval.
Also not sure exactly how to use `undo()`. Simply defining it like `destroy()` doesn't seem to have it called.

In my case I would do something like this to clear the interval:

```js
Alpine.data('counter', () => ({
    counter: 0,
    counter_interval: null,

    init() {
        this.counter_interval = setInterval(() => {
            if ( this.$refs.counter ) {
                this.counter++;
                this.$refs.counter.innerText = this.counter;
            } else {
                clearInterval(this.counter_interval);
            }
        }, 1000);
    },
}));
```

When the interval gets called and the reference isn't defined anymore, I would clear the interval.

What do you think of this? 😅

**Reply by josephmiclaus** [reply]
Aaah! Your solution is quite illuminating! Now I can see, how this all works!

**Reply by SimoTod** [reply]
No worries. P.s. Instead of this.`$refs.counter.innerText` + x-ref you can just use x-text on the element.

**Reply by josephmiclaus** [reply]
Yes! Thanks!

I must confess the example was only hypothetical to show my issue with clearing the interval and making sure nothing stayed in memory. 😃
